src/main/java/com/google/sps/servlets/Util.java

```diff
@@ -89,19 +90,22 @@ public static String getAPIKey() throws IOException {
     return str;
   }
 
-  private static String getRefreshToken(String userId) throws IOException, TooManyResultsException {
+  private static Optional<String> getRefreshToken(String userId) throws IOException, TooManyResultsException {
     Query query = new Query("RefreshToken").setFilter(FilterOperator.EQUAL.of("userId", userId));
     PreparedQuery results = datastore.prepare(query);
     Entity entity = results.asSingleEntity();
     if (entity == null) {
-        throw new IOException("could not get refresh token");
+        return Optional.empty();
     }
     String refreshToken = (String) entity.getProperty("refreshToken");
-    return refreshToken;
+    return Optional.of(refreshToken);
   }
 
   public static String getAccessToken(String userId) throws IOException, TokenResponseException, TooManyResultsException {
-    final String refreshToken = getRefreshToken(userId);
+    final Optional<String> refreshToken = getRefreshToken(userId);
+    if (!refreshToken.isPresent()) {
+        throw new IOException("no refresh token found");
```

Comment:
Hmm... This is just moving the exception around, and not actually addressing "don't throw exception for expected path" concern. But further looking into the code, looks like we are currently redirecting in DevicesServlet.java to AuthorizeServlet when we need to get new refresh token, so throwing an exception seems to be inevitable with current code structure.

Reply:
yep... i do think that it would be possible however to change the exception to be more specific instead of using ioexception as a catch all - for example, using notauthorizedexception to indicate that we should be redirecting to authorize servlet